lib/pbench/test/unit/server/test_file_tree.py

```diff
@@ -251,32 +246,26 @@ def test_unpack_with_returncode(
         """Test to check the unpacking functionality of the FileTree when
         returncode value is not zero"""
         source_tarball, source_md5, md5 = tarball
-        tarball_name = source_tarball.name[0 : -len(".tar.xz")]
 
-        def mock_run(args, cwd, stdin, stderr, stdout):
+        def mock_run(args, cwd, stdin, stderr, stdout, universal_newlines):
             assert args == [
                 "tar",
                 "-xf",
                 (server_config.ARCHIVE / "ABC" / source_tarball.name),
             ]
-            assert cwd == server_config.INCOMING / "ABC" / tarball_name
-            assert stderr == subprocess.DEVNULL
             return subprocess.CompletedProcess(
-                args, returncode=1, stdout=b"Some error unpacking tar", stderr=None
+                args, returncode=1, stdout=None, stderr="Some error unpacking tarball"
```

Comment:
As I mentioned elsewhere, I'm not sure that setting `stderr=STDOUT` will produce this behavior in the real function, so we probably need to correct the behavior of the mock....

Reply:
```
Python 3.9.7 (default, Sep 13 2021, 08:18:39) 
[GCC 8.5.0 20210514 (Red Hat 8.5.0-3)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> p = subprocess.run(["cat", "/foo/bar"], stdin=subprocess.DEVNULL, capture_output=True)
>>> p
CompletedProcess(args=['cat', '/foo/bar'], returncode=1, stdout=b'', stderr=b'cat: /foo/bar: No such file or directory\n')
>>> 
```

The main problem here is that we're getting byte string data; and notice the trailing newline, which the mock is missing. Instead of decoding the byte string, telling `subprocess` to return it as text makes more sense:

```
>>> p = subprocess.run(["cat", "/foo/bar"], stdin=subprocess.DEVNULL, capture_output=True, text=True)
>>> p
CompletedProcess(args=['cat', '/foo/bar'], returncode=1, stdout='', stderr='cat: /foo/bar: No such file or directory\n')
```

Reply:
Your mock behavior is consistent with your unit test validation, so this works. However, as my example shows, the *real* `subprocess.run` will return the trailing newline on the `stderr` message, which you're not including.

I'm undecided on whether this actually matters, since the only real point is to show that `.unpack` is faithfully passing the actual data, whatever is it. Still, since this isn't *my* conversation to close, I'll point out that one trivial "weakness". 😆